src/Bicep.Core/TypeSystem/UnassignableType.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Bicep.Core.Diagnostics;
+
+namespace Bicep.Core.TypeSystem
+{
+    public class UnassignableTypeSymbol : TypeSymbol
+    {
+        private const string ErrorTypeName = "error";
+        private readonly ImmutableArray<ErrorDiagnostic> errors;
+        private readonly TypeKind kind;
+
+        public UnassignableTypeSymbol(string name, TypeKind kind)
+            : base(name)
+        {
+            this.kind = kind;
+
+            // if errors are not initialized, initialize it as an empty array
+            if (this.errors.IsDefault)
+            {
+                this.errors = ImmutableArray<ErrorDiagnostic>.Empty;
+            }
+        }
+
+        public UnassignableTypeSymbol(ErrorDiagnostic error)
+            : this(ErrorTypeName, TypeKind.Error)
+        {
+            this.errors = ImmutableArray.Create<ErrorDiagnostic>(error);
+        }
+
+        public UnassignableTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
+            : this(ErrorTypeName, TypeKind.Error)
+        {
+            this.errors = errors.ToImmutableArray();
```

Comment:
To avoid setting errors in multiple places, one option is to have a constructor that sets everything, and have all other constructors call into it with different args - e.g.:
```
public UnassignableTypeSymbol(string name, TypeKind kind, ImmutableArray<ErrorDiagnostic> errors)
    : base(name)
{
    this.TypeKind = kind;
    this.errors = errors;
}

public UnassignableTypeSymbol(string name, TypeKind kind)
    : this(name, kind, ImmutableArray<ErrorDiagnostic>.Empty)
{
}

public UnassignableTypeSymbol(ErrorDiagnostic error)
    : this(ErrorTypeName, TypeKind.Error, ImmutableArray.Create<ErrorDiagnostic>(error))
{
}

public UnassignableTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
    : this(ErrorTypeName, TypeKind.Error, errors.ToImmutableArray())
{
}
```

The other thing I'm wondering is - is it worth doing something to differentiate the creation of an error type from an arbitrarily-named type, e.g. by using a named static function:
```
UnassignableTypeSymbol.CreateError(errors);
```
vs
```
// implicitly returns an 'error' type
new UnassignableTypeSymbol(errors);
```